segmentation.py

```diff
@@ -0,0 +1,575 @@
+                                           self.__omega,
+                                           self.__sigma,
+                                           self.__eta)
+
+        # Create the initial hypothesis, which states that the first segment is
+        # about to begin.
+        self.__update_count = 0
+        self.__hypotheses = [{'index': 0,
+                              'count': 0,
+                              'log_probability': 0.0,
+                              'distribution': stat,
+                              'log_constant': stat.logconst()}]
+
+    def __accum(self, x, y):
+        return max(x, y) + np.log1p(np.exp(-abs(x - y)))
+
+    def block_update(self, X, Y):
+        """Update model with multiple observations.
+
+        This method is a convenience method for updating the model when many
+        new input-output data are available. This method simply iterates over
+        the inputs calling :py:meth:`.update`.
+
+        Args:
+            X (numpy.array): Observed (K x M) input data (predictor variable).
+            Y (numpy.array): Observed (K x N) output data (response variable).
+
+        """
+
+        # Iterate through the rows of the input-output data updating the
+        # hypotheses in the model.
+        for i in range(X.shape[0]):
+            self.update(X[i, :], Y[i, :])
+
+    def update(self, X, Y):
+        """Update model with a single observation.
+
+        When new input-output data is available, the model can be updated using
+        this method. As more and more data are collected, the number of
+        hypotheses grows, increasing the computational complexity. By default
+        hypotheses are pruned at the end of each update (see
+        :py:meth:`.trim`.). To disable hypotheses trimming, initialise the
+        class with ``maxhypot`` set to ``None``.
+
+        Args:
+            X (numpy.array): Observed (1 x M) input data (predictor variable).
+            Y (numpy.array): Observed (1 x N) output data (response variable).
+
+        """
+
+        # Initialise algorithm on first call to update. This allows the
+        # algorithm to configure itself to the size of the first input/output
+        # data if no hyper-parameters have been specified.
+        if not self.__initialised:
+            m = X.shape[1] if np.ndim(X) > 1 else X.size
+            n = Y.shape[1] if np.ndim(Y) > 1 else Y.size
+            self.__initialise_algorithm(m, n)
+            self.__initialised = True
+
+        # Get size of data.
+        k = X.shape[0] if np.ndim(X) > 1 else X.size
+        m, n = self.__m, self.__n
+
+        # Allocate variables for the dynamic programming pass.
+        loglik = -np.inf
+        logmax = -np.inf
+        logsum = -np.inf
+        ind = np.nan
+
+        # Update hypotheses by updating each maxtrix variate, normal inverse
+        # gamma distribution over the linear models.
+        self.__update_count += 1
+        for hypotheses in self.__hypotheses:
```

Comment:
`for hypothesis in self.__hypotheses:` we're iterating over hypothes**e**s, one hypothes**i**s at a time.